File: whoareyou.rb
#!/usr/bin/ruby 

# Based on whoareyou.py

require 'digest/md5'
require 'base64'


def determine_identity()
	#puts 'starting determine_identity'
	agent_fingerprints = []
	IO.popen("ssh-add -l") {|line|
	  l = line.gets
	  keylen, fingerprint, rest = l.split(' ', 3)
	  agent_fingerprints << fingerprint.gsub(/:/,"")
  }
  # Read authorized_keys file
  File.open(File.expand_path("~/.ssh/authorized_keys")).each do |line|
    keytype, blob, comment = line.split(' ', 3)
    fingerprint = Digest::MD5.hexdigest(Base64.decode64(blob))
    #agent_fingerprints << fingerprint
    if agent_fingerprints.include?(fingerprint) then
      return comment
    end
  end
  return "<unknown>"


end

puts determine_identity

File: whoareyou.py
#!/usr/bin/python

# Find out who is running this script over SSH, by:
# 1. Asking ssh-agent for their known key fingerprints
# 2. Looking for a matching key in .ssh/authorized_keys, and reporting the key
#        comment.

import binascii
import os
import warnings
warnings.filterwarnings('ignore', '[Tt]he (md5|popen2) module is deprecated', DeprecationWarning)
import md5
import popen2

def determine_identity():
    agent_fingerprints = []
    ssh_add = popen2.Popen4(['ssh-add', '-l'])
    for line in ssh_add.fromchild.readlines():
        try:
            keylength, fingerprint, remainder = line.split(' ', 2)
        except ValueError:
            pass
        agent_fingerprints.append(fingerprint.replace(':', ''))
    try:
        for line in open(os.path.expanduser('~/.ssh/authorized_keys')):
            try:
                keytype, blob, comment = line.strip().split(' ', 2)
                fingerprint = md5.md5(binascii.a2b_base64(blob)).hexdigest()
            except (ValueError, binascii.Error):
                continue
            if fingerprint in agent_fingerprints:
                return comment
    except IOError:
        # no file, we ignore
        pass
    return "<unknown>"

if __name__ == '__main__':
    print determine_identity()
